abci/types/types.go

```diff
@@ -117,54 +117,72 @@ func (r *EventAttribute) MarshalJSON() ([]byte, error) {
-// the conversion, make that a non-issue.
+// UnmarshalJSON was modified in the Celestia fork to be backwards compatible
+// with the EventAttribute from CometBFT v0.34.x. CometBFT v0.38.x uses the type
+// string for keys and values. CometBFT v0.34.x used the type bytes for keys and
+// values. CometBFT v0.34.x event attributes that were marshaled to JSON
+// previously encoded the keys and values as base64 strings so this method
+// attempts to base64 decode the keys and values if they look like base64
+// encoded data.
 func (r *EventAttribute) UnmarshalJSON(b []byte) error {
-	// Parse the JSON into a raw struct to inspect the format
-	var raw struct {
+	var eventAttribute struct {
 		Key   string `json:"key,omitempty"`
 		Value string `json:"value,omitempty"`
 		Index bool   `json:"index,omitempty"`
 	}
 
-	if err := json.Unmarshal(b, &raw); err != nil {
+	if err := json.Unmarshal(b, &eventAttribute); err != nil {
 		return err
 	}
 
-	// Try to decode Key as base64, fallback to raw string
-	if raw.Key != "" {
-		if keyBytes, err := base64.StdEncoding.DecodeString(raw.Key); err == nil {
-			// Check if the base64 decoded content looks like it should be decoded
-			// (i.e., it's not the same as the original and produces valid output)
-			if string(keyBytes) != raw.Key {
-				r.Key = string(keyBytes)
-			} else {
-				r.Key = raw.Key
-			}
-		} else {
-			r.Key = raw.Key
+	r.Key = maybeBase64Decode(eventAttribute.Key)
+	r.Value = maybeBase64Decode(eventAttribute.Value)
+	r.Index = eventAttribute.Index
+	return nil
+}
+
+func maybeBase64Decode(input string) string {
+	if isLikelyBase64Encoded(input) {
+		bytes, err := base64.StdEncoding.DecodeString(input)
+		if err != nil {
+			return input // input is not a base64 encoded string so return the input
 		}
+		decoded := string(bytes)
+		return decoded
 	}
+	return input
+}
 
-	// Try to decode Value as base64, fallback to raw string
-	if raw.Value != "" {
-		if valueBytes, err := base64.StdEncoding.DecodeString(raw.Value); err == nil {
-			// Check if the base64 decoded content looks like it should be decoded
-			// (i.e., it's not the same as the original and produces valid output)
-			if string(valueBytes) != raw.Value {
-				r.Value = string(valueBytes)
-			} else {
-				r.Value = raw.Value
-			}
-		} else {
-			r.Value = raw.Value
+// isLikelyBase64Encoded returns true if input is likely a base64 encoded string.
+func isLikelyBase64Encoded(input string) bool {
+	bytes, err := base64.StdEncoding.DecodeString(input)
+	if err != nil {
+		return false
+	}
+	decoded := string(bytes)
+	if input == decoded {
+		return false
+	}
+
+	// Only decode if the result is printable ASCII/UTF-8 text
+	for _, r := range decoded {
+		// Allow printable ASCII characters, spaces, and common unicode
+		if r < 32 && r != '\t' && r != '\n' && r != '\r' {
+			return false
+		}
+		// Reject high-value bytes that are likely binary garbage
+		if r > 127 && r == '\ufffd' {
+			return false
```

Comment:
This looks to be too complicated to maintain and could be a source of eventual errors. Any way to simplify it? what about adding a field in core in the response that would specify whether these are base64 encoded or not? or define an activation height where the events before are handled as base64, the ones after as strings? 

I'm trying to think of a different way to do this without having to guess if it's base64. `aGloaQ` is valid as a string and as base64 and with guessing, we won't give users an API that always gives valid results

Reply:
fwiw we can get rid of this code eventually, this is just an annoying compatibility thing with v3 and v4

I don't know if there is a perfect way to do this

Reply:
Agreed difficult to maintain and it is guessing but that's also true of the original fix PR https://github.com/celestiaorg/celestia-core/pull/2184

If we don't have a 100% reliable way to support backwards compatibility for v0.34.x events then maybe we don't apply the patch in `UnmarshalJSON` and try to fix it some other way.